### ISSUE TITLE
A `}` or `|` inside an AMDP / EXEC SQL body hides ENDMETHOD / ENDEXEC, and the following statements are not parsed

### ISSUE THREAD

**Issue by oisee** [abaplint/abaplint#4329]
Hi Lars,

when the body of a `METHOD ... BY DATABASE PROCEDURE|FUNCTION ...` contains a `}` or a `|` outside an ABAP string -- including inside a SQLScript comment (`-- ...` or `/* ... */`) -- the rest of the file up to the next `|` or `{` is lexed as one token per line, so `ENDMETHOD.` is no longer a statement, and neither is the next `METHOD ... BY DATABASE ...`. `)`, `]`, a lone `{` and `'}'` in a string do not trigger it. The same happens in `EXEC SQL ... ENDEXEC.`

Reproducer (@abaplint/core 2.120.55 and 2.120.59):

```abap
CLASS zcl_r DEFINITION PUBLIC.
  PUBLIC SECTION.
    INTERFACES if_amdp_marker_hdb.
    CLASS-METHODS a.
    CLASS-METHODS b.
ENDCLASS.
CLASS zcl_r IMPLEMENTATION.
  METHOD a BY DATABASE PROCEDURE FOR HDB LANGUAGE SQLSCRIPT.
    declare x integer;
    x = 1; -- closing }
  ENDMETHOD.
  METHOD b BY DATABASE PROCEDURE FOR HDB LANGUAGE SQLSCRIPT.
    declare y integer;
  ENDMETHOD.
ENDCLASS.
```

```js
const reg = new abaplint.Registry().addFile(new abaplint.MemoryFile("zcl_r.clas.abap", src)).parse();
reg.getFirstObject().getABAPFiles()[0].getStatements().map((s) => s.get().constructor.name);
```

gives

`ClassDefinition Public InterfaceDef MethodDef MethodDef EndClass ClassImplementation MethodImplementation NativeSQL`

and the last tokens are `Identifier "ENDMETHOD."`, `Identifier "METHOD b BY DATABASE PROCEDURE FOR HDB LANGUAGE SQLSCRIPT."`, ... ; `structure` reports "Expected ENDMETHOD". With `-- closing )` instead it is `... MethodImplementation NativeSQL EndMethod MethodImplementation NativeSQL EndMethod EndClass`, as expected.

For the kernel the body is SQLScript and ends at `ENDMETHOD.`; classes with a `}` in an AMDP comment activate on a system.

It looks like the lexer's `ModeNormal` branch that starts a string template on `|` / `}` (`lexer.ts`, "start template"): the template mode is not ended by a newline, and the native-SQL handling in the statement parser only re-wraps statements after lexing. Just a pointer, I have not tried a fix.

We cut AMDP bodies out of the text as a workaround, so this is not urgent for us. Possibly related: #3486, #4307.
